python/amici/sbml_import.py

```diff
@@ -989,6 +998,67 @@ def replace_assignments(formula: str) -> sp.Basic:
         self.symbols['llhy']['name'] = l2s(llh_y_syms)
         self.symbols['llhy']['identifier'] = llh_y_syms
 
+    def _process_conservation_laws(self, ode_model) -> None:
+        """
+        Find conservation laws in reactions and species.
+        """
+
+        conservation_laws = []
+        # So far, only conservation laws for constant species are supported
+        self._add_conservation_for_constant_species(ode_model,
+                                                    conservation_laws)
+
+        # add the found CLs to the ode_model
+        for cl in conservation_laws:
+            ode_model.add_conservation_law(**cl)
+
+    def _add_conservation_for_constant_species(
+            self,
+            ode_model: ODEModel,
+            conservation_laws: List[ConservationLaw]
+    ) -> None:
+        """
+        Adds constant species to conservations laws
+
+        """
+
+        # decide which species to keep in stoichiometry
+        species_solver = list(range(ode_model.nx_rdata()))
+        S = self.stoichiometric_matrix
+
+        for ix in range(ode_model.nx_rdata()):
+            if ode_model.state_is_constant(ix):
+                target_state = ode_model._states[ix]._identifier
+                total_abundance = ode_model._states[ix].get_val()
+                conservation_laws.append({
+                    'state': target_state,
+                    'total_abundance': total_abundance,
+                    'state_expr': total_abundance,
+                    'abundance_expr': target_state,
+                })
+                self._replace_in_all_expressions(target_state, total_abundance)
```

Comment:
Isn't this automatically handled by the ConservationLaw implementation?

Reply:
Not 100% sure, but I think not. It didn't work for me without this line... But I can retry...

Reply:
Looks like it wasn't necessary in https://github.com/ICB-DCM/AMICI/blob/390305f98c96ea505cf52884991774e7dfefdeff/python/amici/pysb_import.py#L359

Reply:
douvble-checked, it is necessary.
pysb_import is constructed somewhat differently, but also calls a similar functionality in _flatten_conservation_laws. There, also substitutions are called...
So, I could either use replace_in_all_expressions, or copy-paste flatten_conservation_laws and adapt it, if necessary...
If you prefer the latter, I can doe the latter. But the current solution does the job without code duplication, so I used this one.

Reply:
but `_flatten_conservation_laws` only operates on the conservation laws to remove interdependencies and doesn't modify any other expressions.

I think this implementation is conceptually not correct. The right way to go is to use `target_state` as `state_expr`. What is currently being done is hardcopying the total abundance everywhere where the state occurs. You may get away with that, but it's not in the spirit of the conservation law implementation and might prevent conservation laws from correctly updating after reinitialization with updated fixedParameters (may be worthwhile to add a corresponding test).

Reply:
Although I have to double check since its implemented the same way in `_add_conservation_for_constant_species` 🤔

Reply:
Okay I think the issue is more about the way total abundance is specified here. I don't think you want to pass the value here, but rather a new symbolic variable such as `total_abundance = sp.Symbol(f'tcl_{target_state}')

Reply:
The point is also a bit that I didn't totally get where to put "target_state" and where "state_expression". I tried different things, none of which worked without the "replace in all expression". I feel like state_expression should actually be something totally different, neither target_state nor state_expression, but rather an independent symbolic...

I felt like it's kinda difficult to get those four lines correct because I don't fully understand what is finally replaced where in add_conservation_law. It's pretty hard to develop an intuition for that, as the overall idea, how the code should be affected, is somewhat missing for me... :/

Reply:
> Okay I think the issue is more about the way total abundance is specified here. I don't think you want to pass the value here, but rather a new symbolic variable such as `total_abundance = sp.Symbol(f'tcl_{target_state}')`

I think I even tried that one without success... 😅 
I'd be really thankful for some help on those lines, as I feel a little lost there...
I can try your suggestion, but I think something would still be missing...

Reply:
Sounds like I need to add more documentation to `add_conservation_law`, will do.

Reply:
yep... The problem is that the old state-expression is *not* replaced in the flux vector by the current code. This causes that the old state expression still shows up in cpp-files, which don't know the identifier any more, resulting finally in a compiler error.
I somehow need to replace the old state expression in the flux vector, otherwise it seems like it won't work.
I tried quite some permutations and adding other symbolic expressions now...

Reply:
Added documentation. I think the key issue that currently isn't addressed is that `add_conservation_law` adds an Expression to the model to compute the value of the conserved quantity. This means that `ode_model._eqs['w'] = self.flux_vector` no longer works but that flux_vector should be translated into expressions.

Reply:
Okay, the doc helps, that's nice!
However, the problem persists, that the flux_vector needs to be changed...
Would a simple `subs(state, state_expr)` thrown on the flux vector do the trick? Or would something be missing?

Reply:
And... we shouldn't forget: Things may look slightly different for real conservation laws, but this is actually only the interface for constant species. This code isn't even meant to capture actual conserved quantities across states...

Reply:
Ok, I changed the substitution. This is now the minimal version it needs to work.
Ich you want, I can still add a symbolic idnetifier in here, but this should anyway be a symbolic substitution...

Reply:
I don't think the current implementation works with reinitialization of states/fixedParameters. I think the right way to go is to is

- [ ] pass the values as described above to `add_conservation_law`
- [ ] remove assignment to `ode_model._eqs['w']`, instead use `add_component` to add flux vector entries to the ode_model
- [ ] store the symbols used in `add_component` and reconsitute the flux vector (with symbolic identifiers), compute `xdot` based on those and update `dxdotdw` (add a couple zero rows in the right place).

Reply:
> I don't think the current implementation works with reinitialization of states/fixedParameters. I think the right way to go is to is

But my unit test uses this reinitialization and works... (?)

Reply:
Okay then it probably works in most cases. Looking at the generated code, the part that is circumvented is there to avoid issues with larger expressions for conservation laws that get repeatedly differentiated and thereby cause and explosion of the size of derivatives.

Still not happy with the current implementation as it explicitly doesn't use the API as intended and will inevitably confuse the next person looking at that code.

Reply:
okay, then I'll change it. Code is supposed to stay maintainable... 👍

Reply:
Update on this one:
I will need to include the constants k into total_cl.cpp, as the conservation laws now need to be informed about constant parameters. Otherwise, conservation_laws depending on constant parameters won't work...

Reply:
Some more questions to this...

> * [ ]  remove assignment to `ode_model._eqs['w']`, instead use `add_component` to add flux vector entries to the ode_model

Shall I add them as generic expressions, or introduce a new class "flux"?

> * [ ]  store the symbols used in `add_component` and reconsitute the flux vector (with symbolic identifiers), compute `xdot` based on those and update `dxdotdw` (add a couple zero rows in the right place).

Wait... This means that xdot is not S * w any more, and that would mean that I have to alter the stoichiometric matrix S, beyond just deleting the zero-rows, right?
I totally see the purpose: It would introduce the framework for implementing real conservation laws directly and I would already do half the job which is necessary to implement those (except computing the left kernel of S and identifying the most promising species to be replaced).

Two things on this:
Firstly: this is substantially more changing than I was planning to do right now. Not crucially problematic, but wanted to point it out. This was never the aim of this pull-request.
Secondly, and substantially more important: I really think this isn't the best way to go for species, which are just initialized with super-simple rules. It adds overhead to the symbolic processing (don't know how much, but definitely not few), by refactoring the whole stoichiometry, although this will never be necessary for this very specific type of conservation law. For hits reason, I really think it is better to keep simple things simple and not refactor fluxes and the stoichiometry, if we know that we don't need to...

Reply:
No, just use the `add_component(Expression(...))`, keep track of the symbolic identifiers for those expression and add them to a vector `r`. You can then calculate `xdot` as `S*r` instead of `S*w`, only `dxdotdw` needs additional rows for non-flux Expressions (all of them come from the conservation laws. Shouldn't be too much changes.

Reply:
i.e.: `r` is then basically `w`, just with the additional expressions, one per CL?
`xdot = S * r` then?